src/cmd/ksh93/tests/emacs.exp

```diff
@@ -0,0 +1,279 @@
+    puts "ctrl-n works"
+}
+
+# Take cursor at beginning of line with ctrl-a
+send "\x01"
+expect -exact "" {
+    puts "ctrl-a works"
+}
+
+# Take cursor to the end of line with ctrl-e
+send "\x05"
+expect -exact "set -o emacs" {
+    puts "ctrl-e works"
+}
+
+# Move cursor backward one character with ctrl-b
+send "\x02"
+expect -exact "\x08" {
+    puts "ctrl-b works"
+}
+
+# Move cursor forward one character with ctrl-f
+# First move to the beginning of line
+send "\x01"
+expect -exact ""
+send "\x06"
+expect -exact "s" {
+    puts "ctrl-f works"
+}
+
+# Move cursor forward one word with alt-f
+send "\033f"
+expect -exact "et" {
+    puts "alt-f works"
+}
+
+#
+#
+# ^X^X ?
+#
+
+# Delete a character with ctrl-d
+# Move to the beginning of line
+send "\x01"
+send "\x04"
+expect -exact "et -o emacs" {
+    puts "ctrl-d deletes a char"
+}
+
+# Delete current word with alt-d
+send "\033d"
+expect -exact "-o emacs" {
+    puts "alt-d deletes a word"
+}
+
+# Take cursor to the end of line with ctrl-e
+send "\x05"
+# Delete previous word with ALT-backspace
+send "\033\x08"
+expect -exact "-o" {
+    puts "alt-backspace deletes previous word"
+}
+
+# Clear the commandline with ctrl-c
+send "\x03"
+expect -exact ""
+
+# Set command and move to end of line
+send "echo hello"
+send "\x05"
+
+# Delete previous word with ALT-h
+send "\033h"
+expect -exact "echo" {
+    puts "alt-h deletes previous word"
+}
+
+# Delete previous word with alt-delete
+send "\033\x7F"
+expect -exact "" {
+    puts "alt-delete deletes previous word"
+}
+
+# Transpose character with ctrl-t
+send "echo hello\x14"
+expect -exact "echo helol" {
+    puts "ctrl-t transposes characters"
+}
+
+# Clear command line
+send "\x03"
+send "echo hello"
+# Move to beginning of line
+send "\x01"
+# Capitalize current word with alt-c
+# This should capitalize the word 'echo' to 'ECHO'
+send "\033c"
+expect -exact "ECHO" {
+    puts "alt-c capitalizes current word"
+}
+
+# Change current word to lowercase alt-l
+# Move to beginning of line
+send "\x01"
+send "\033l"
+expect -exact "echo" {
+    puts "alt-l changes current word to lowercase"
+}
+
+# Move to beginning of line
+send "\x01"
+# Delete from cursor to end of line with ctrl-k
+send "\x11"
+# Cursor is at beginning of line so that should delete full line
+expect -exact "" {
+    puts "ctrl-k deletes characters to end of line"
+}
+
+# Kill from cursor to mark with ctrl-w
+send "echo hello2"
+send "\x17"
+expect -exact "" {
+    puts "ctrl-w kills from cursor to mark"
+}
+
+#
+# M-p       Push the region from the cursor to the mark on the stack.
+#
+# ^Y        Restore last item removed from line. (Yank item back to the line.)
+send "\x19"
+expect -exact "echo hello2" {
+    puts "ctrl-y yanks back item to the line"
+}
+
+#
+# ^L        Line feed and print current line
+#
+# M-^L      Clear the screen
+#
+
+# ^J        (New line) Execute the current line.
+#
+# ^M        (Return) Execute the current line.
+send "\x03"
```

Comment:
In general every test should ensure a new prompt is displayed and accounted for when it is done. That means every test should end with an `expect_prompt` call to ensure that is the case. We shouldn't be using [ctrl-C] to force a fresh prompt.

Reply:
Would it work if command on prompt is not executed ?

Reply:
The idea is to ensure we keep in sync with what's happening by inserting a `expect_prompt` everyplace we expect a new prompt to be issued. That way if we get out of sync, either because an unexpected prompt appears or the one we expect does not appear, we find out ASAP. So if your test is doing something like erasing the current char/word/line then we don't expect a new prompt so `expect_prompt` should not be called. But something like [ctrl-C] should cause a new prompt so we should verify that actually occurs.

Reply:
Also, I don't yet understand why line 255 is failing but when it does the input buffer looks like this:

```
[DEBUG] timeout; buffer= \r\nKSH PROMPT:15: ^[^\x08
```

Notice the escape char followed by a literal carat, then a backspace char.

Reply:
Does macOS handle `alt-ctrl-v` differently ?

Reply:
No. The very handy `fish_key_reader` program (part of the fish shell) shows the expected output:

```
Press a key

              hex:   1B  char: \c[  (or \e)
(  0.225 ms)  hex:   16  char: \cV
bind \e\cV 'do something'
```

And it works fine when I type that sequence at an interactive shell prompt. There is something about the unit test that seems to cause the escape produced by [meta] to not be tied to the following [ctrl-V]. Which causes the [ctrl-V] to have it's usual "quote the next special char" meaning. So, yes, superficially this looks like a timing issue but that should be impossible given how the test is written.

If I delete all the tests prior to the [meta-ctrl-V] test it then works reliably.

Reply:
This seems to have something to do with how [ctrl-C] is handled.